### PR TITLE
fix(cache): harden routing and cache lifecycle

### PR DESCRIPTION
**Pull request by Gajesh2007**
## Summary

- closes every validated post-merge cache-routing review finding across the coordinator and Swift provider
- pins engine hardening from Layr-Labs/mlx-swift-lm#73
- keeps cache routing and early prompt donation disabled by default

### Behavior

```mermaid
flowchart LR
  A[Consumer request] --> B[Coordinator dispatch]
  B --> C{Provider protocol}
  C -->|v0| D[Caller-derived or shared cache namespace]
  C -->|v1| E[Cache route without mandatory provider hash match]
  D --> F[Provider inference]
  E --> F
  F --> G[SSD staging reserves shared headroom]
  G -->|staging refuses| H[Capacity error even when cold request fits]
  F --> I[Cache hit latency enters cold prefill and TTFT calibration]
  F --> J[Terminal message can overtake or omit lookup receipt]
  F --> K[Raw provider cache details can escape in streamed payloads]
```

### Code

```mermaid
flowchart LR
  A[registry cache evidence] --> B[scheduler discount]
  B --> C[api dispatch]
  C --> D[ProviderLoop inference handler]
  D --> E[EngineV2Bridge]
  E --> F[SSDPrefixCache path-based I/O]
  D --> G[independent receipt and terminal sends]
  D --> H[prefill EWMA]
  C --> I[TTFT and reputation samples]
```

### Behavior

```mermaid
flowchart LR
  A[Consumer request] --> B[Coordinator creates isolated attempt]
  B --> C{Remote cache eligible?}
  C -->|protocol v0 or routing off| D[Fresh per-attempt cache buster]
  C -->|v1 and exact provider/catalog hash match| E[Scoped cache route]
  C -->|no hash match| F[Cold unscoped-disabled request]
  E --> G[Provider performs pre/post cryptographic load-hash bracket]
  G -->|verified match| H[Memory or encrypted SSD cache]
  G -->|missing or changed| F
  H --> I{SSD staging fits?}
  I -->|yes| J[Stage and serve]
  I -->|no| K[Release staging and retry full cold reservation]
  J --> L[Receipt queued before exactly one terminal]
  K --> L
  F --> L
  L --> M[Sanitized usage and correlated terminal telemetry]
  M --> N[Cache participants excluded from cold-prefill calibration]
```

### Code

```mermaid
flowchart LR
  A[cache_receipts.go] --> B[cache_watermarks.go]
  B --> C[cache_routing.go and scheduler.go]
  C --> D[api dispatch.go immutable attempt identity]
  D --> E[ProviderLoop InferenceHandler and ModelLoading]
  E --> F[PrefixCacheReceipts ordered finalizer]
  E --> G[EngineV2Bridge request-aware receipt identity]
  G --> H[SSDPrefixCache namespaced tickets]
  H --> I[SSDNoFollowIO descriptor-relative operations]
  G --> J[mlx-swift-lm donation lifecycle hardening]
  F --> K[api provider.go idempotent outcome metrics]
  K --> L[consumer.go cache-detail sanitization]
```

## Details

- isolates protocol-0 and routing-off requests from reusable remote cache namespaces
- requires matching catalog and provider weight hashes before remote cache participation
- requires matching pre-load and post-load cryptographic hashes before reusable SSD cache construction, including standalone mode
- prevents stale evidence resurrection with bounded watermarks protected by active-attempt heaps
- includes `X-Session-Id` in exact route identity and rejects non-finite routing configuration
- retries cold capacity admission after optional SSD staging abandonment
- binds staging tickets to request correlation IDs and namespaces process-wide reservations per cache instance
- performs active Darwin SSD reads, writes, touches, renames, and deletes through descriptor-relative no-follow operations
- finalizes exactly one lookup receipt before terminal completion or error across outer and bridge failure paths
- sanitizes raw cache-token details from complete, Responses, content, finish, escaped-key, and multiline SSE payloads
- joins selection and terminal telemetry without recording client identifiers or cache keys
- excludes cache-participating latency from provider cold-prefill EWMA and coordinator TTFT/reputation calibration
- binds early donations to their source state and includes donation-owned state in graceful drain readiness; paged early donation remains disabled

## Verification

- `cd coordinator && go test ./...`
- `cd coordinator && go test -race ./registry`
- final focused coordinator API, registry, store, dispatch, sanitization, calibration, and telemetry tests passed
- `cd provider-swift && swift test --parallel` passed 1,309 tests in 145 suites
- liveness-recovery and whole-root-maintenance fixtures pass in CI parallel mode without shared-budget or asynchronous-start races
- final provider receipt/hash selection passed 8/8
- final SSD selection passed 57/57
- engine early-donation selection passed 16/16
- engine prefix-cache selection passed 37/37
- engine scheduler, end-to-end, multimodal, and quantized selection passed 119/119
- final integrated engine selection passed 72/72
- the full engine suite built successfully but exceeded the 30-minute runner timeout without failure output; focused affected suites passed
- `git diff --check`

## Dependency

- engine PR: Layr-Labs/mlx-swift-lm#73
- parent gitlink: `89aa0eeab`

## Rollout

No release or deploy is included. Cache routing and early prompt donation remain default-off.

### CODE REVIEW

File: coordinator/api/consumer.go [hunk 1]
```diff
@@ -1645,6 +1677,7 @@ func (s *Server) handleStreamingResponseWithFirstChunk(w http.ResponseWriter, r
 		if firstChunk == "" || isSSEDoneChunk(firstChunk) {
 			continue
 		}
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:1679]: 🔵 **[INFO]** ⚡ sanitizeStreamCacheDetails called on every chunk in streaming response

💡 _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant JSON parsing

📊 _Score: 2×3 = 6 · Category: repeated_work_

File: coordinator/api/consumer.go [hunk 2]
```diff
@@ -2479,21 +2514,180 @@ func injectReasoningDetailIntoRawUsage(obj map[string]any, usage protocol.UsageI
...
+	// A normal field takes the literal fast path. A disguised equivalent must use
+	// a JSON Unicode escape (for example cached\u005ftokens); no other valid JSON
+	// escape can encode an ASCII letter or underscore. Parse those frames too,
+	// while leaving the overwhelmingly common escape-free content frame untouched.
+	if !strings.Contains(chunk, `"cached_tokens"`) && !strings.Contains(chunk, `\u`) {
+		return chunk
+	}
+	normalized := strings.ReplaceAll(strings.ReplaceAll(chunk, "\r\n", "\n"), "\r", "\n")
+	groups := strings.Split(normalized, "\n\n")
+	changed := false
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:2680]: 🟡 **[MEDIUM]** ⚡ sanitizeStreamCacheDetails performs O(n) string operations and JSON parsing on every streaming chunk

💡 _Suggestion:_ Optimize by checking for presence of cache tokens before expensive JSON parsing, or use streaming JSON parser for large chunks

📊 _Score: 3×4 = 12 · Category: inefficient_data_structures_

File: coordinator/registry/cache_watermarks.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+// tombstone churn cannot evict ordering evidence needed by a delayed receipt.
+func (t *cacheRoutingTracker) enforceWatermarkCapLocked() {
+	for len(t.watermarks) > t.maxWatermarks {
+		if len(t.watermarkOrder) == 0 {
+			// Defensive only: the 2*maxAttempts sizing invariant makes this
+			// unreachable while attempt and route-key caps hold.
+			return
+		}
+		t.removeWatermarkLocked(t.watermarkOrder[0].ref)
+	}
```
Comment (ethenotethan) [inline coordinator/registry/cache_watermarks.go:172]: 🟡 **[MEDIUM]** ⚡ enforceWatermarkCapLocked uses heap operations in a loop without batching

💡 _Suggestion:_ Consider batching watermark removals or using a more efficient bulk removal strategy when many watermarks need eviction

📊 _Score: 3×4 = 12 · Category: inefficient_data_structures_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                }
+                guard next >= 0 else {
+                    throw SSDBlockStoreError.ioFailure(
+                        "openat directory \(component) under \(canonicalPath): "
+                            + String(cString: strerror(errno)))
+                }
+                Darwin.close(current)
+                current = next
+            }
+            return current
```
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:210]: 🟡 **[MEDIUM]** ⚡ openDirectoryChain opens and closes file descriptors for each path component

💡 _Suggestion:_ Cache opened directory descriptors for frequently accessed paths or use a more efficient path traversal method

📊 _Score: 3×3 = 9 · Category: repeated_work_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — ✅ No issues found

### Performance — 4 finding(s) (3 blocking)
- 🔵 **[INFO]** `coordinator/api/consumer.go:1677-1679` — sanitizeStreamCacheDetails called on every chunk in streaming response
  - _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant JSON parsing
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:2514-2680` — sanitizeStreamCacheDetails performs O(n) string operations and JSON parsing on every streaming chunk
  - _Suggestion:_ Optimize by checking for presence of cache tokens before expensive JSON parsing, or use streaming JSON parser for large chunks
- 🟡 **[MEDIUM]** `coordinator/registry/cache_watermarks.go:162-172` — enforceWatermarkCapLocked uses heap operations in a loop without batching
  - _Suggestion:_ Consider batching watermark removals or using a more efficient bulk removal strategy when many watermarks need eviction
- 🟡 **[MEDIUM]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:185-210` — openDirectoryChain opens and closes file descriptors for each path component
  - _Suggestion:_ Cache opened directory descriptors for frequently accessed paths or use a more efficient path traversal method

### Type_diligence — ✅ No issues found

### Additive_complexity — ✅ No issues found

_4 finding(s) total, 3 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        // component before the O_NOFOLLOW descriptor walk.
+        let canonicalPath: String
+        if rawPath == "/var" || rawPath.hasPrefix("/var/") {
+            canonicalPath = "/private" + rawPath
+        } else if rawPath == "/tmp" || rawPath.hasPrefix("/tmp/") {
+            canonicalPath = "/private" + rawPath
+        } else {
+            canonicalPath = rawPath
+        }
+        guard canonicalPath.hasPrefix("/") else {
```
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:190]: 🔵 **[INFO]** 🔒 Hardcoded system path normalization may be incomplete

💡 _Suggestion:_ Consider using realpath() or equivalent system call for complete path resolution instead of textual prefix replacement

📊 _Score: 3×2 = 6 · Category: path-traversal_

File: coordinator/api/consumer.go [hunk 6]
```diff
@@ -2479,21 +2514,180 @@ func injectReasoningDetailIntoRawUsage(obj map[string]any, usage protocol.UsageI
...
+
+func removeCachedTokenDetail(usageObj map[string]any, detailField string) bool {
+	details, ok := usageObj[detailField].(map[string]any)
+	if !ok || details == nil {
+		return false
+	}
+	if _, exists := details["cached_tokens"]; !exists {
+		return false
+	}
+	delete(details, "cached_tokens")
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:2580]: 🔵 **[INFO]** 🔒 Complex cache detail sanitization logic may have edge cases

💡 _Suggestion:_ Add comprehensive unit tests for all sanitization paths and consider using a allowlist approach for trusted fields

📊 _Score: 2×3 = 6 · Category: input-validation_

File: coordinator/api/consumer.go [hunk 7]
```diff
@@ -1645,6 +1677,7 @@ func (s *Server) handleStreamingResponseWithFirstChunk(w http.ResponseWriter, r
 		if firstChunk == "" || isSSEDoneChunk(firstChunk) {
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:1677]: 🟡 **[MEDIUM]** ⚡ sanitizeStreamCacheDetails called on every chunk in hot streaming path

💡 _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk

📊 _Score: 2×4 = 8 · Category: repeated_work_

File: coordinator/api/consumer.go (hunk 2 above)
Comment (ethenotethan) [inline coordinator/api/consumer.go:2680]: 🟡 **[MEDIUM]** ⚡ sanitizeStreamCacheDetails performs O(n) string operations on every streaming frame

💡 _Suggestion:_ Use more efficient string processing or pre-compile regex patterns for the common case where no sanitization is needed

📊 _Score: 3×3 = 9 · Category: inefficient_data_structures_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift (hunk 4 above)
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:210]: 🔵 **[INFO]** ⚡ openDirectoryChain performs synchronous directory operations in a loop

💡 _Suggestion:_ Consider batching directory operations or using async I/O where possible to avoid blocking on each path component

📊 _Score: 2×3 = 6 · Category: blocking_io_

File: coordinator/api/consumer.go [hunk 8]
```diff
@@ -866,24 +866,32 @@ func (s *Server) dispatchOneProvider(
...
+		s.registry.ForgetCacheAttempt(pr)
+		refundExtra()
+		cleanupPending()
+		return nil, nil, decision, "failed to prepare cache-safe request", http.StatusInternalServerError
+	}
 	// Pre-fix providers crash on a vision request carrying sampling penalties;
-	// strip them for those providers only (see bodyForProvider).
-	sealedBody := bodyForProvider(rawBody, requiresVision, provider)
+	// strip them for those providers only. Protocol-0 providers additionally get
+	// a coordinator-authored prompt_cache_key only inside this sealed body.
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:877]: 🟡 **[MEDIUM]** 🧩 PrepareCacheAttempt called before bodyForCacheAttempt creates temporal coupling

💡 _Suggestion:_ Consider combining these operations or making the dependency explicit through the function signature

📊 _Score: 3×4 = 12 · Category: over-abstraction_

File: coordinator/api/consumer.go [hunk 9]
```diff
@@ -965,6 +973,30 @@ func bodyForProvider(rawBody []byte, requiresVision bool, provider *registry.Pro
...
+	sealed, err := marshalForwardBody(parsed)
+	if err != nil {
+		return nil, err
+	}
+	if len(sealed) > maxInferenceBodyBytes {
+		return nil, fmt.Errorf("provider request body exceeds the %d-byte limit after cache isolation", maxInferenceBodyBytes)
+	}
+	return sealed, nil
+}
+
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:999]: 🟡 **[MEDIUM]** 🧩 bodyForCacheAttempt duplicates bodyForProvider logic with additional cache-specific handling

💡 _Suggestion:_ Extract common body transformation logic to avoid duplication between bodyForProvider and bodyForCacheAttempt

📊 _Score: 4×3 = 12 · Category: over-abstraction_

File: coordinator/api/consumer.go [hunk 10]
```diff
@@ -2479,21 +2514,180 @@ func injectReasoningDetailIntoRawUsage(obj map[string]any, usage protocol.UsageI
...
 	}
-	usageObj["prompt_tokens_details"] = details
+}
+
+func removeCachedTokenDetail(usageObj map[string]any, detailField string) bool {
+	details, ok := usageObj[detailField].(map[string]any)
+	if !ok || details == nil {
+		return false
+	}
+	if _, exists := details["cached_tokens"]; !exists {
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:2577]: 🟡 **[MEDIUM]** 🧩 Multiple similar cache detail sanitization functions with overlapping logic

💡 _Suggestion:_ Consolidate sanitizeCacheDetailIntoRawResponsesUsage, injectCacheDetailIntoRawUsage, and removeCachedTokenDetail into a unified cache detail handler

📊 _Score: 3×3 = 9 · Category: over-abstraction_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        at url: URL,
+        modificationDate: Date,
+        beforeOperation: (@Sendable (SSDActiveIOOperation) -> Void)? = nil
+    ) {
+        beforeOperation?(.touch)
+        try? FileManager.default.setAttributes(
+            [.modificationDate: modificationDate], ofItemAtPath: url.path)
+    }
+    #endif
+}
```
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:293]: 🔴 **[CRITICAL]** 🧩 Entire new file implementing custom file I/O operations that duplicate standard library functionality

💡 _Suggestion:_ Evaluate if the security requirements truly necessitate custom I/O implementation or if existing secure file handling patterns could be used

📊 _Score: 5×3 = 15 · Category: reinventing-stdlib_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — 2 finding(s)
- 🔵 **[INFO]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:178-190` — Hardcoded system path normalization may be incomplete
  - _Suggestion:_ Consider using realpath() or equivalent system call for complete path resolution instead of textual prefix replacement
- 🔵 **[INFO]** `coordinator/api/consumer.go:2514-2580` — Complex cache detail sanitization logic may have edge cases
  - _Suggestion:_ Add comprehensive unit tests for all sanitization paths and consider using a allowlist approach for trusted fields

### Performance — 4 finding(s) (3 blocking)
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:1677` — sanitizeStreamCacheDetails called on every chunk in hot streaming path
  - _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:2514-2680` — sanitizeStreamCacheDetails performs O(n) string operations on every streaming frame
  - _Suggestion:_ Use more efficient string processing or pre-compile regex patterns for the common case where no sanitization is needed
- 🟡 **[MEDIUM]** `coordinator/registry/cache_receipts.go:214-290` — applyLookup iterates through all route keys multiple times with nested loops
  - _Suggestion:_ Optimize the key iteration pattern or use more efficient data structures to reduce the O(n²) behavior
- 🔵 **[INFO]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:175-210` — openDirectoryChain performs synchronous directory operations in a loop
  - _Suggestion:_ Consider batching directory operations or using async I/O where possible to avoid blocking on each path component

### Type_diligence — ✅ No issues found

### Additive_complexity — 5 finding(s) (5 blocking)
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:869-877` — PrepareCacheAttempt called before bodyForCacheAttempt creates temporal coupling
  - _Suggestion:_ Consider combining these operations or making the dependency explicit through the function signature
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:976-999` — bodyForCacheAttempt duplicates bodyForProvider logic with additional cache-specific handling
  - _Suggestion:_ Extract common body transformation logic to avoid duplication between bodyForProvider and bodyForCacheAttempt
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:2517-2577` — Multiple similar cache detail sanitization functions with overlapping logic
  - _Suggestion:_ Consolidate sanitizeCacheDetailIntoRawResponsesUsage, injectCacheDetailIntoRawUsage, and removeCachedTokenDetail into a unified cache detail handler
- 🔴 **[CRITICAL]** `coordinator/api/consumer.go:2579-2726` — sanitizeStreamCacheDetails implements complex SSE parsing that reinvents parts of standard SSE handling
  - _Suggestion:_ Consider using or extending existing SSE parsing libraries rather than implementing custom SSE field parsing
- 🔴 **[CRITICAL]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:1-293` — Entire new file implementing custom file I/O operations that duplicate standard library functionality
  - _Suggestion:_ Evaluate if the security requirements truly necessitate custom I/O implementation or if existing secure file handling patterns could be used

_11 finding(s) total, 8 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_

File: coordinator/api/consumer.go (hunk 8 above)
Comment (ethenotethan) [reply coordinator/api/consumer.go:877]: :white_check_mark: Looks addressed in `f150f1d` — PrepareCacheAttempt is now called before bodyForCacheAttempt and the dependency is made explicit through the function signature. Resolving.

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift (hunk 5 above)
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:190]: 🔵 **[INFO]** 🔒 Hardcoded system path normalization may be incomplete

💡 _Suggestion:_ Consider using realpath() or equivalent system call for complete path resolution instead of textual prefix replacement

📊 _Score: 3×2 = 6 · Category: path-traversal_

File: coordinator/api/consumer.go (hunk 6 above)
Comment (ethenotethan) [inline coordinator/api/consumer.go:2580]: 🔵 **[INFO]** 🔒 Complex cache detail sanitization logic may have edge cases

💡 _Suggestion:_ Add comprehensive unit tests for all sanitization paths and consider using a allowlist approach for trusted fields

📊 _Score: 2×3 = 6 · Category: input-validation_

File: coordinator/api/consumer.go [hunk 12]
```diff
@@ -1645,6 +1677,7 @@ func (s *Server) handleStreamingResponseWithFirstChunk(w http.ResponseWriter, r
 		if firstChunk == "" || isSSEDoneChunk(firstChunk) {
 			continue
 		}
+		firstChunk = sanitizeStreamCacheDetails(firstChunk)
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:1680]: 🔵 **[INFO]** ⚡ sanitizeStreamCacheDetails called on every chunk in hot streaming path

💡 _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant JSON parsing

📊 _Score: 2×3 = 6 · Category: repeated_work_

File: coordinator/api/consumer.go [hunk 13]
```diff
@@ -2479,21 +2514,180 @@ func injectReasoningDetailIntoRawUsage(obj map[string]any, usage protocol.UsageI
...
+	if json.Unmarshal([]byte(raw), &obj) != nil {
+		return raw, false
+	}
+	changed := false
+	if usageObj, ok := obj["usage"].(map[string]any); ok {
+		changed = removeCachedTokenDetail(usageObj, "prompt_tokens_details") || changed
+		changed = removeCachedTokenDetail(usageObj, "input_tokens_details") || changed
+	}
+	if response, ok := obj["response"].(map[string]any); ok {
+		if usageObj, ok := response["usage"].(map[string]any); ok {
```
Comment (ethenotethan) [inline coordinator/api/consumer.go:2600]: 🟡 **[MEDIUM]** ⚡ sanitizeStreamCacheDetails performs expensive string operations and JSON parsing on every streaming frame

💡 _Suggestion:_ Add early exit conditions (e.g., check for presence of cached_tokens or \u before expensive operations) and consider streaming-specific optimizations

📊 _Score: 3×4 = 12 · Category: repeated_work_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift (hunk 4 above)
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:210]: 🔵 **[INFO]** ⚡ openDirectoryChain opens and closes file descriptors for each path component

💡 _Suggestion:_ Consider caching opened directory descriptors for frequently accessed paths or batch directory operations where possible

📊 _Score: 2×3 = 6 · Category: repeated_work_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — 2 finding(s)
- 🔵 **[INFO]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:178-190` — Hardcoded system path normalization may be incomplete
  - _Suggestion:_ Consider using realpath() or equivalent system call for complete path resolution instead of textual prefix replacement
- 🔵 **[INFO]** `coordinator/api/consumer.go:2514-2580` — Complex cache detail sanitization logic may have edge cases
  - _Suggestion:_ Add comprehensive unit tests for all sanitization paths and consider using a allowlist approach for trusted fields

### Performance — 4 finding(s) (1 blocking)
- 🔵 **[INFO]** `coordinator/api/consumer.go:1680` — sanitizeStreamCacheDetails called on every chunk in hot streaming path
  - _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant JSON parsing
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:2514-2600` — sanitizeStreamCacheDetails performs expensive string operations and JSON parsing on every streaming frame
  - _Suggestion:_ Add early exit conditions (e.g., check for presence of cached_tokens or \u before expensive operations) and consider streaming-specific optimizations
- 🔵 **[INFO]** `coordinator/registry/cache_receipts.go:214-290` — Nested loops over cache route keys without early termination
  - _Suggestion:_ Add break statements when watermark sequence conditions fail early to avoid unnecessary iterations through remaining keys
- 🔵 **[INFO]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:180-210` — openDirectoryChain opens and closes file descriptors for each path component
  - _Suggestion:_ Consider caching opened directory descriptors for frequently accessed paths or batch directory operations where possible

### Type_diligence — ✅ No issues found

### Additive_complexity — ✅ No issues found

_6 finding(s) total, 1 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_

File: coordinator/api/consumer.go (hunk 1 above)
Comment (ethenotethan) [inline coordinator/api/consumer.go:1679]: 🔵 **[INFO]** ⚡ sanitizeStreamCacheDetails called on every chunk in streaming response

💡 _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant string operations

📊 _Score: 2×3 = 6 · Category: repeated_work_

File: coordinator/api/consumer.go (hunk 2 above)
Comment (ethenotethan) [inline coordinator/api/consumer.go:2680]: 🟡 **[MEDIUM]** ⚡ sanitizeStreamCacheDetails performs O(n) string operations on every streaming chunk

💡 _Suggestion:_ Optimize the sanitization by using a single pass parser or pre-checking for the presence of cached_tokens before expensive string operations

📊 _Score: 3×4 = 12 · Category: inefficient_data_structures_

File: coordinator/registry/cache_watermarks.go (hunk 3 above)
Comment (ethenotethan) [inline coordinator/registry/cache_watermarks.go:172]: 🟡 **[MEDIUM]** ⚡ enforceWatermarkCapLocked uses heap operations in a loop without batching

💡 _Suggestion:_ Consider batching watermark removals or using a more efficient eviction strategy to avoid O(log n) heap operations per eviction

📊 _Score: 3×4 = 12 · Category: inefficient_data_structures_

File: provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift (hunk 4 above)
Comment (ethenotethan) [inline provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:210]: 🔵 **[INFO]** ⚡ openDirectoryChain performs synchronous directory traversal with multiple syscalls

💡 _Suggestion:_ Consider caching opened directory descriptors or using async I/O for the directory chain traversal to avoid blocking on filesystem operations

📊 _Score: 2×3 = 6 · Category: blocking_io_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — 1 finding(s)
- 🔵 **[INFO]** `coordinator/api/consumer.go:2479` — JSON unmarshaling without size limits on user-controlled cache body
  - _Suggestion:_ Add size validation before json.Unmarshal in bodyForCacheAttempt to prevent potential DoS via large JSON payloads

### Performance — 4 finding(s) (2 blocking)
- 🔵 **[INFO]** `coordinator/api/consumer.go:1677-1679` — sanitizeStreamCacheDetails called on every chunk in streaming response
  - _Suggestion:_ Consider caching the sanitization result or checking if sanitization is needed before processing each chunk to avoid redundant string operations
- 🟡 **[MEDIUM]** `coordinator/api/consumer.go:2514-2680` — sanitizeStreamCacheDetails performs O(n) string operations on every streaming chunk
  - _Suggestion:_ Optimize the sanitization by using a single pass parser or pre-checking for the presence of cached_tokens before expensive string operations
- 🟡 **[MEDIUM]** `coordinator/registry/cache_watermarks.go:162-172` — enforceWatermarkCapLocked uses heap operations in a loop without batching
  - _Suggestion:_ Consider batching watermark removals or using a more efficient eviction strategy to avoid O(log n) heap operations per eviction
- 🔵 **[INFO]** `provider-swift/Sources/ProviderCore/KVCacheSSD/SSDNoFollowIO.swift:175-210` — openDirectoryChain performs synchronous directory traversal with multiple syscalls
  - _Suggestion:_ Consider caching opened directory descriptors or using async I/O for the directory chain traversal to avoid blocking on filesystem operations

### Type_diligence — ✅ No issues found

### Additive_complexity — ✅ No issues found

_5 finding(s) total, 2 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_
